File: cmd/utils/downloadFile.ts
import { createWriteStream, unlinkSync } from "fs";
import http from "http";
import https from "https";
import ensureDir from "./ensureDir";
import progressBar from "./progressBar";

function formatBytes(bytes: number) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    while (bytes >= 1024) {
        bytes /= 1024;
        units.shift()
    }
    return {
        unit: units[0],
        value: bytes.toFixed(2)
    }
}

/**
 * Download a file over the internet and store it on the local machine
 */
export default function downloadFile(fileUrl: string, outputPath: string) {
    return new Promise((resolve, reject) => {
        const parsedUrl = new URL(fileUrl);
        const protocol = parsedUrl.protocol === "https:" ? https : http

        ensureDir(outputPath)
        const file = createWriteStream(outputPath)


        protocol
            .get(fileUrl, (response: http.IncomingMessage) => {

                response.pipe(file)

                file
                    .on("finish", () => {
                        file.close()
                        resolve(true)
                    })
                    .on("error", (err: Error) => {
                        file.close()
                        reject("Pipe Error: " + err.message)
                    })

                const totalSize = parseInt(response.headers['content-length'] || '0', 10);
                let downloadedSize = 0
                const bar = progressBar();

                response.on("data", (chunk: Buffer) => {
                    downloadedSize += chunk.length
                    bar({
                        download: formatBytes(downloadedSize),
                        source: formatBytes(totalSize)
                    })
                })

            })
            .on("error", err => {
                unlinkSync(outputPath)
                reject("Download Error: " + err.message)
            })
    })
}

File: go/utils/download_file.go
package utils

import (
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
)

// download data at a url to a particular path on local computer
func DownloadFile(url string, path string) error {
	fmt.Printf("Downloading %s to %s \n", url, path)
	
	out, err := os.Create(path)
	if err != nil {
		return err
	}
	defer out.Close()
	
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()	

	if resp.StatusCode != http.StatusOK {
		return errors.New("Download failed!")
	}

	_, err = io.Copy(out, resp.Body)
	if err != nil {
		return err
	}

	return nil
}